cmd/crypto/vault.go

```diff
@@ -0,0 +1,149 @@
+package crypto
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"strconv"
+
+	vault "github.com/hashicorp/vault/api"
+	"github.com/minio/minio/cmd/logger"
+)
+
+const (
+	// VaultEndpointEnv Vault endpoint environment variable
+	VaultEndpointEnv = "MINIO_VAULT_ENDPOINT"
+	// VaultAppRoleIDEnv  Vault App Role ID environment variable
+	VaultAppRoleIDEnv = "MINIO_VAULT_APP_ROLE_ID"
+	// VaultAppSecretIDEnv Vault App Secret ID environment variable
+	VaultAppSecretIDEnv = "MINIO_VAULT_APP_SECRET_ID"
+	// VaultAppVersionEnv Vault App Version environment variable
+	VaultAppVersionEnv = "MINIO_VAULT_APP_VERSION"
+	// VaultKeyNameEnv Vault Encryption Key Name environment variable
+	VaultKeyNameEnv = "MINIO_VAULT_KEYNAME"
+)
+
+type VaultService struct {
+	config *VaultConfig
+	client *vault.Client
+}
+
+func (v *VaultService) encryptEndpoint(key string) string {
+	return "/transit/encrypt/" + key
+}
+func (v *VaultService) decryptEndpoint(key string) string {
+	return "/transit/decrypt/" + key
+}
+
+// VaultConfig holds config required to start vault service
+type VaultConfig struct {
+	Endpoint      string
+	AppRoleID     string
+	AppSecretID   string
+	AppVersion    string
+	EncryptionKey string // should this be defined by client / generated on the fly
+}
+
+// validate whether all required env variables needed to start vault service have
+// been set
+func validateVaultConfig(c *VaultConfig) error {
+	if c.Endpoint == "" {
+		return fmt.Errorf("Invalid value set in environment variable %s", VaultEndpointEnv)
+	}
+	if c.AppRoleID == "" {
+		return fmt.Errorf("Invalid value set in environment variable %s", VaultAppRoleIDEnv)
+	}
+	if c.AppSecretID == "" {
+		return fmt.Errorf("Invalid value set in environment variable %s", VaultAppSecretIDEnv)
+	}
+	if c.AppVersion != "" {
+		_, err := strconv.Atoi(c.AppVersion)
+		if err != nil {
+			return fmt.Errorf("Invalid value set in environment variable %s", VaultAppVersionEnv)
+		}
+	}
+	if c.EncryptionKey == "" {
```

Comment:
Actually you're right my bad. For vault the Key-ID must be non-emtpy and must exist on the vault because of the reason you mentioned. For the SSE-S3 master key variant it is fine, but in this case the empty-string check is required.